### ISSUE TITLE
iOS/Android: Text input error for screenreaders

### ISSUE THREAD

**Issue by amarlette** [facebook/react-native#30848]
## Requires API Proposal
This issue requires a new API. An API proposal should be added and discussed before proceeding with implementation. The API proposal can be added in the comments of this issue or linked as a separate issue.

## Description
Cannot specify error text on TextInputs. Errors for form fields, while they can be displayed visually in many unique ways, need to have a strong association with the problematic field itself for accessibility. Otherwise, any users who can't visually see the relation of the error to the field won't necessarily know which field is the problem.

## React Native version:
v0.63

## Expected Results
Upon focus of a text input in an error state, the error message specified should announced after the main focus announcement, prefixed with "Error: ". Upon changing the contents of a text input which puts the input into an error state, an error message should be announced prefixed with "Error: ".

## Snack
https://snack.expo.io/X4NIqnJ2K

## Android Details
To implement error states on Android, you should call the `setError` method of the `TextView`.  This will automatically set the correct properties on the AccessibilityNodeInfo that will inform screen readers to this state. If you need to do this manually, what this does behind the scenes is call `setContentInvalid(true)` and `setError(youErrorString)` on the AccessibilityNodeInfo.

## iOS Details
iOS has no standard pattern for the presentation of error states on text inputs, so I'd recommend we follow Android's pattern here.  To do this, you'll need to implement a key press listener, and on change, detect if the field is now in an error state.  If it is, you'll need to make a manual screen reader announcement of the error message, and append this message to the accessibilityValue for the field.

**Comment by kacieb** [comment]
# API Proposal:

This will require adding a new prop to JS, and Android and iOS changes to use that prop.

JS

Add a new prop to TextInput:
```
  /**
   * String to be read by screenreaders to indicate an error state. If this value is
   * not null, an error will be announced. You can use onChangeText or onBlur to 
   * detect an error and set this prop. Once the error is gone, set this to null
   * to clear the error
   */
  errorMessage?: ?Stringish,
```
Example usage:
In this example, the TextInput goes into an error state if the user types "error" into the TextInput.
```
export default function Playground(): React.Node {
  const [text, setText] = React.useState('');
  const [error, setError] = React.useState(null);
  return (
    <TextInput
      error={error}
      onChangeText={newText => {
        setText(newText);
        setError(newText === 'error' ? 'this input is invalid' : null);
      }}
      value={text}
    />
  );
}
```

Note: `onChangeText` may not be the right place to announce the error. This API gives flexibility to set the error when needed. Another place might be `onBlur`, which is when the TextInput loses focus.

# Implementation Thoughts
On Android, implementation will follow the suggestions in the issue. To avoid visual styling, we will call setContentInvalid(true) and setError(youErrorString) on the AccessibilityNodeInfo manually.

On iOS, since there is no way to do this sanctioned by Apple, we'll follow the recommended approach listed in the issue. The behavior will be as similar to Android as possible.

**Comment by zersiax** [comment]
Curious to know why this was moved back onto the backlog? This is definitely something that would be a dealbreaker for me if it woldn't work as described here.

**Comment by kacieb** [comment]
> Curious to know why this was moved back onto the backlog? This is definitely something that would be a dealbreaker for me if it woldn't work as described here.

Hello! I changed teams and no longer had capacity to work on it, which is why I released the issue. It'd be really great if you wanted to pick this up, though! You can follow my proposal above if you're interested in working on this.

**Comment by fabriziobertoglio1987** [comment]
@blavalla I would like to work on this after completing the review of https://github.com/facebook/react-native/pull/31666 Thanks :pray:

**Comment by fabriziobertoglio1987** [comment]
Thanks a lot @blavalla 

1- Issues iOS/Android: [Text input error for screenreaders #30848][1] and [Android: Error state change (text inputs) #30859][2] should be solved together (as explained in Brett [comment][3]).
2- Issue [Android: Error state change (text inputs) #30859][2] may require updates to Android API (see test case [Change of error state from onChangeText show/hides a TextInput error][5]). I found a draft solution for the issue and I think is related to [updateState][10]. I will prepare a fix for this issue in a couple of days.
3- I tested Kacie's [API Proposal][8] (videos of my [tests][6], my draft [branch][7]). I agree with her API Proposal. The prop name would be `android_errorMessage`.
4- I did not yet investigate the iOS API

[1]: https://github.com/facebook/react-native/issues/30848
[2]: https://github.com/facebook/react-native/issues/30859
[3]: https://github.com/facebook/react-native/issues/30859#issuecomment-786970372
[4]: https://github.com/fabriziobertoglio1987/react-native/blob/3b7581cd763635aad605cc175bc3cb74182f4a85/ReactAndroid/src/main/java/com/facebook/react/views/textinput/ReactTextInputManager.java#L1328 "ReactTextUpdate"
[5]: https://github.com/fabriziobertoglio1987/react-native-notes/issues/12#issuecomment-1071989570 "test case"
[6]: https://github.com/fabriziobertoglio1987/react-native-notes/issues/12#issuecomment-1072101477
[7]: https://github.com/facebook/react-native/compare/main...fabriziobertoglio1987:text-input-errors-build
[8]: https://github.com/facebook/react-native/issues/30848#issuecomment-801449852
[9]: https://github.com/facebook/react-native/pull/29070
[10]: https://github.com/fabriziobertoglio1987/react-native/blob/3b7581cd763635aad605cc175bc3cb74182f4a85/ReactAndroid/src/main/java/com/facebook/react/views/textinput/ReactTextInputManager.java#L1328-L1334 "TextInputManager updateState()"

**Comment by fabriziobertoglio1987** [comment]
I added the following test cases. I tested the PR on Fabric and Paper iOS.

[76]. iOS - announce lastChar (not entire text) onChangeText and avoid multiple announcements (Fabric) ([link][76])
[77]. iOS - announces or does not announce the accessibilityError through Button onPress (not onChangeText) (Fabric) ([link][77])
[78]. iOS - the error is announced with accessibilityInvalid true and does not clear after typing text (onChangeText) (Fabric) ([link][78])

I'll complete the below TODOs and move the PR to ready to Review by Monday morning (New York time).

- [ ] Implement changes from my own [code review](https://github.com/facebook/react-native/pull/33468#pullrequestreview-1009172601)
- [ ] Implement solution for multiline TextInput (iOS)
- [ ] Further review, test and improve the pull request


[76]: https://github.com/facebook/react-native/issues/30859#issuecomment-1158790381 "iOS - announce lastChar (not entire text) onChangeText and avoid multiple announcements (Fabric)"
[77]: https://github.com/facebook/react-native/issues/30859#issuecomment-1158794863 "iOS - announces or does not announce the accessibilityError through Button onPress (not onChangeText) (Fabric)"
[78]: https://github.com/facebook/react-native/issues/30859#issuecomment-1158797801 "iOS - the error is announced with accessibilityInvalid true and does not clear after typing text (onChangeText) (Fabric)"

**Comment by fabriziobertoglio1987** [comment]
<details><summary>iOS - Exception thrown while executing UI block: - RCTUITextView setAccessibilityErrorMessage:]: unrecognized selector sent to instance (iOS - Paper on main branch)</summary>
<p>


<image src="https://user-images.githubusercontent.com/24992535/173279577-98f1e7e2-dda8-4c25-b834-cb8b7933a522.jpeg" width="300" />

</p>
</details>

**Comment by fabriziobertoglio1987** [comment]
I'll move the PR as Ready for Review, once I complete these tasks:

- [ ] fix regression in Fabric iOS. Reintroduce solution from commit https://github.com/facebook/react-native/pull/33468/commits/318a96420f0593dc59c030193009e9cba31eb666
- [ ] extensive iOS and Android tests
- [ ] review diff and PR summary

**Comment by fabriziobertoglio1987** [comment]
I moved PR https://github.com/facebook/react-native/pull/33468 Ready for Review. 
I remain available for any improvement. Thanks

**Comment by fabriziobertoglio1987** [comment]
PR not merged, could add it as separate library

**Comment by zersiax** [comment]
This is considered basic functionality. This should not be a library, as developers will not know of its existence and therefore render an accessibility experience that is inferior without the package installed. This should be part of core.

**Comment by fabriziobertoglio1987** [comment]
I'm not talking of a single fix, but creating a new clone of the [TextInput component](https://github.com/fabriziobertoglio1987/react-native-improved) outside of react-native.

I have been contributing to facebook/react-native for 4 years full-time and I have 60 PRs. I'm also the SE that published the PR for this issue but did not make it to main.

I would keep maintaining and improving the library. Users would have the choice between TextInput component from facebook/react-native and TextInputImproved component from fabriziobertoglio1987/react-native-improved

Do you prefer to have at least a choice in this situation?

If you had no choice (at least until the PRs are merged, which may take years), would you use TextInput from an external library that includes 100-1000 fixes for TextInput and all the original functionalities? 

I’m asking because I’m organizing a list of issues and I will start working on them soon. Thanks

**Comment by zersiax** [comment]
My problem with that approach is that developers aren't going to make that choice. They won't know the choice exists, or they won't care enough to make the choice, or their hands will be tied by management/an overzealous product owner/what have you to not make the choice for all sorts of different reasons. It'd be one more thing to maintain, one more third-party dependency. 
Proper error prevention and reporting is among the basics of basics where accessibility is concerned. Not getting this right is a failure of WCAG, which will get a company sued and will give users who actually need this to work a bad user experience. Apps on mobile are already being yanked under this same umbrella and it's up to the maintainers of React Native to realize that this has priority and can't be hanging out in the wings for years. If you reckon this will take years, I'm all for creating a fork of the component so there is at least an option for devs who care about not discriminating against a bunch of users. This really needs to be in main though, eventually.
